pypeit/spectrographs/keck_kcwi.py

```diff
@@ -1031,46 +1032,164 @@ def get_rawimage(self, raw_file, det):
+        the remained comes back through the optical system and a fuzzy version of this is re-imaged onto the
+        detector. The current KCWI scattered light model is designed to account for these effects.
 
         Parameters
         ----------
         frame : `numpy.ndarray`_
             Raw 2D data frame to be used to compute the scattered light.
-        binning : str, `numpy.ndarray`_, tuple
-            Binning of the frame (e.g. '2x1' refers to a binning of 2 in the spectral
-            direction, and a binning of 1 in the spatial direction). For the supported
-            formats, refer to :func:`~pypeit.core.parse.parse_binning`.
+        offslitmask : `numpy.ndarray`_
+            A boolean mask indicating the pixels that are on/off the slit (True = off the slit)
+        tbl : :class:`~pypeit.metadata.PypeItMetaData`_
+            One row of the fitstbl PypeItMetaData that contains metadata about the file being used to
+            optimize the scattered light model parameters.
+        detpad : :obj:`int`_, optional
+            Number of pixels to pad to each of the detector edges to reduce edge effects.
 
         Returns
         -------
         scatt_img : `numpy.ndarray`_
             A 2D image of the scattered light determined from the input frame
         """
-        # Determine the spatial binning. Currently, we hard-code the regions on the KCWI detector
-        # to be used for the determination of the scattered light. This is crude, and a better model
-        # is currently under development (RJC)
-        spatbin = parse.parse_binning(binning)[1]
-        ym = 4096 // (2 * spatbin)
-        y0 = ym - 180 // spatbin
-        y1 = ym + 180 // spatbin
-        # Obtain a robust (median) "spectrum" of the scattered light at the left, right, and middle of the detector
-        scattlightl = np.nanmedian(frame[:, :20//spatbin], axis=1)[:, None]
-        scattlightr = np.nanmedian(frame[:, -40//spatbin:], axis=1)[:, None]
-        scattlight = np.nanmedian(frame[:, y0:y1], axis=1)[:, None]
-        # The following algorithm is a polynomial fit to:
-        # (1) The left half of the detector (using the left and middle scattered light spectrum)
-        # (2) The right half of the detector (using the right and middle scattered light spectrum)
-        # We then ensure that the model is continuous and smooth at the middle of the detector.
-        medl = np.median(scattlightl / scattlight)
-        medr = np.median(scattlightr / scattlight)
-        spatimg = np.meshgrid(np.arange(frame.shape[1]), np.arange(frame.shape[0]))[0]
-        scatt_img = np.outer(scattlight, np.ones(frame.shape[1]))
-        scatt_img[spatimg < ym] *= 1 + (medl - 1) * (spatimg[spatimg < ym] / ym - 1) ** 2
-        scatt_img[spatimg >= ym] *= 1 + (medr - 1) * (spatimg[spatimg >= ym] / (4095//spatbin - ym) - ym / (4095//spatbin - ym)) ** 2
-        return scatt_img
+        def resid(param, wpix, img):
```

Comment:
Nested (Inner) functions are a pet peeve of mine, in the sense that I'd rather we not do this, but I realize this is largely a style preference.

Reply:
I also hesitate about this, but it's not that much different to the lambda function (I think), and this is a bit clearer, because we can add a docstring. I'm not really sure how to do this otherwise (i.e. call a spectrograph specific model).